### PR TITLE
fix: terminate the process correctly

### LINKED ISSUE #2026: `fetch` can make `worker_threads` stuck and process hang
## Bug Description

When `undici` is run inside `worker_threads` it may prevent the worker from terminating and makes the whole process hang.

## Reproducible By

This Node script runs `undici` in worker threads and does 4x requests there before exiting. If this does not reproduce on your fast environment, try increasing the requests and `const ROUNDS`. 

```
"undici": "^5.21.0"
```

<details>

```js
// Save as `undici.mjs` and run as `node uncidi.mjs`
import { isMainThread, parentPort, Worker } from "node:worker_threads";
import { fileURLToPath } from "node:url";
import { cpus } from "node:os";

const THREADS = cpus().length - 1;
const ROUNDS = 10;
const filename = fileURLToPath(import.meta.url);

if (isMainThread) {
  const stuckWorkers = [];
  const interval = setInterval(() => {
    if (stuckWorkers.length > 0) {
      console.log(`There are ${stuckWorkers.length} stuck workers`);
    }
  }, 1000);

  const timeout = setTimeout(() => {
    setTimeout(() => console.log("Still running"), 5_000); // This is never called so process.exit does something, but it's still unable to terminate worker running undici
    console.log("Forcing process.exit on main thread");
    process.exit();
  }, 30_000);

  const task = async () => {
    const worker = new Worker(filename, {});

    return new Promise((resolve) => {
      worker.on("message", (message) => {
        if (message !== "DONE") return;

        const timer = setTimeout(() => {
          console.log("Unable to terminate");
          stuckWorkers.push(worker);
        }, 10_000);
        worker.terminate().then(() => {
          clearTimeout(timer);
          resolve();
        });
      });
    });
  };

  const pool = Array(ROUNDS).fill(task);

  async function execute() {
    const task = pool.shift();

    if (task) {
      await task();
      return execute();
    }
  }

  await Promise.all(
    Array(THREADS)
      .fill(execute)
      .map((task) => task())
  );
  console.log("All done!");
  clearTimeout(timeout);
  clearInterval(interval);
} else {
  const { fetch } = await import("undici");

  const request = async (pathWithParams) => {
    const url = new URL(pathWithParams, "https://api.spacexdata.com");

    const timer = setTimeout(
      () => console.log("Request taking more than 10s"), // This is never logged so requests are not stuck
      10_000
    );
    const output = await fetch(url.href, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });
    clearTimeout(timer);
    return output;
  };
  await Promise.all([
    request("/v4/starlink").then((r) => r.json()),
    request("/v4/starlink/5eed770f096e59000698560d").then((r) => r.json()),
    request("/v4/starlink").then((r) => r.json()),
    request("/v4/starlink/5eed770f096e59000698560d").then((r) => r.json()),
  ]);

  parentPort.postMessage("DONE");
}
```

</details>

## Expected Behavior

Package should be safe to run inside `worker_threads`. If it's not safe, it should detect such environment and throw an error.

## Logs & Screenshots

```sh
ari ~/repro  $ node undici.mjs 
Unable to terminate
There are 1 stuck workers
...
There are 1 stuck workers
Forcing process.exit on main thread
^Z
[1]+  Stopped                 node undici.mjs

ari ~/repro  $ pgrep node
24498

ari ~/repro  $ kill  24498
ari ~/repro  $ pgrep node
24498
# ^^ Ok this one is really stuck

ari ~/repro  $ kill  -9 24498
[1]+  Killed: 9               node undici.mjs

ari ~/repro  $ pgrep node
# no output, process killed
```

## Environment

MacOS 13.2.1
Node 18.14.0

### Additional context

https://github.com/vitest-dev/vitest/issues/3077

### LINKED ISSUE #2413: Tests not terminating after execution with msw when using FormData and Blob
## Bug Description

I'm encountering an issue where the process does not terminate after running tests with `msw` when using `FormData` and `Blob`. Removing the line where the `Blob` is appended to the `FormData` allows the tests to terminate as expected.

## Reproducible By

Below is a simplified test case to illustrate the issue:

```
~/work/msw-formdata-issue/ npm test

> msw-formdata-issue@1.0.0 test
> mocha



  test
{}
    ✔ test


  1 passing (22ms)
```

When the above code is executed, the process does not terminate after running the test. However, if the line with `formData.append(...)` is commented out, the process terminates correctly. It seems that the use of FormData in combination with Blob in msw setup is causing some resources to not be released properly.

## Expected Behavior

exit the process after running all tests.

## Environment

Node.js 21.1.0

### Additional context

At first, I noted this issue to msw repo, but msw author suggest to post this issue to undici.
ref. https://github.com/mswjs/msw/issues/1843

### PR DESCRIPTION
**Pull request by tsctx**
Fixes #2413

### CODE REVIEW

**Comment by tsctx** [comment]
Sorry, but I could not create a unit test.

File: lib/fetch/body.js [hunk 1]
```diff
@@ -277,7 +277,8 @@ function cloneBody (body) {
   const out2Clone = structuredClone(out2, { transfer: [out2] })
   // This, for whatever reasons, unrefs out2Clone which allows
   // the process to exit by itself.
-  const [, finalClone] = out2Clone.tee()
+  const [finalClone, final2Clone] = out2Clone.tee()
+  final2Clone.cancel()
```
Comment (KhafraDev) [inline lib/fetch/body.js:281]: this promise isn't being handled

**Review by KhafraDev — COMMENTED** [review]
A test that just makes a fetch call in a worker sounds useful. FWIW I couldn't repro it locally either, but maybe it is in the CIs.

Is there a reason you're using the first tee'd stream instead of the second? IIRC I had issues using the first one.

**Comment by tsctx** [comment]
> A test that just makes a fetch call in a worker sounds useful. FWIW I couldn't repro it locally either, but maybe it is in the CIs.
> 
> Is there a reason you're using the first tee'd stream instead of the second? IIRC I had issues using the first one.

No particular reason. What problems did you have?

**Comment by KhafraDev** [comment]
Don't recall exactly, probably related to the comment

**Comment by tsctx** [comment]
Is this? https://github.com/nodejs/undici/pull/1700

**Comment by KhafraDev** [comment]
that's why we needed the tee, not why I used the second stream. Just wondering, how do you know that this fixes the issue?

**Comment by tsctx** [comment]
I found that I could run a reproducible script of 2 issues and fix it.

**Comment by KhafraDev** [comment]
why can't that be made into a test?

**Comment by tsctx** [comment]
> why can't that be made into a test?

This is because it requires access to an external server.
That test did not work when I set up a local server.

If you want the test, you need a few days. Still, you may not be able to create the test.

**Comment by tsctx** [comment]
My meaning of did not work is that it worked correctly even before the modification.

> as I mentioned before, it might not work locally but it might run on the CI, which is better than nothing (even if it doesn't work on the CI either)

I can't do that right now. I will do some research

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by AriPerkkio** [comment]
@tsctx does this fix reproduction case from #2026 for you locally? I can still reproduce it after applying the changes from this PR.

**Comment by tsctx** [comment]
Hi @AriPerkkio,
Changes are added. Is this fixed?

**Comment by ronag** [comment]
@KhafraDev I'd like you approval on this before merging.

**Comment by tsctx** [comment]
> > Hi @AriPerkkio,
> 
> > Changes are added. Is this fixed?
> 
> 
> 
> No, #2026 still gets stuck with these changes applied locally:
> 
> 
> 
> ```sh
> 
> ari ~/Git/undici (main) $ node repro.mjs 
> 
> Unable to terminate
> 
> ...
> 
> There are 1 stuck workers
> 
> Forcing process.exit on main thread
> 
> ```
> 
> 

Thanks for the report! I thought I had fixed it.

**Comment by KhafraDev** [comment]
@tsctx it's most likely a bug in node, not really sure how much of it can be fixed here. If you find a fix for https://github.com/nodejs/node/issues/44985 it would probably help.

**Pull request closed without merging**